Fix format tutorial on structural boxes

edwintorok opened the pull request:
I was trying to use them as the format tutorial described, but it wasn't working. My initial thoughts were to use break hints with a negative indent, but surely there is a better way.
I'm not the only one who's confused by this, see the discussion at https://discuss.ocaml.org/t/format-module-from-the-standard-library/2254, which led me to https://hal.science/hal-01503081/file/format-unraveled.pdf.
Figure 2b) in that paper shows the solution: use 0 indentation structural boxes, but you need 2 of each, so that the closing parenthesis ends up in the right place.

To be fair the tutorial never said to use indented boxes, OTOH it also didn't say that you need to use 2, just one.
Keep the existing format string in the tutorial, but show what the real output is. And then show how to fix it with the additional structural boxes.
The commit messages contain sample code to show both outputs (with actual format strings and break hints, replacing `[` and `b` from the tutorial)

Finally, this is very error-prone (as evidenced by the tutorial itself being wrong, or at best ambiguous), and show how to write some helper functions (I'll probably propose including something similar into `Fmt`).

edwintorok committed:
doc(format): fix structural box Format tutorial

The reference output was off by 1 space.

Signed-off-by: Edwin Török <edwin@etorok.net>

edwintorok committed:
Update changelog

Signed-off-by: Edwin Török <edwin@etorok.net>

Octachron commented:
I am not convinced that using break hints with a negative indentation is not a better solution in the context of the tutorial. Even simpler updating the reference output to match the real output might be better.

In particular, your code is only equivalent to the one in the "tutorial" for an indentation of `1`, since you are switching
```ocaml
"@[<b 1>(---@,)@]"
```
to
```ocaml
"@[(@[---@]@,)@]"
```
Whereas adding a negative indent for the final space
```ocaml
"@[<b 1>(---@;<0 -1>)@]"
```
works for any indentation.

Octachron commented on api_docgen/Format_tutorial.mld:
```diff
@@ -256,6 +270,23 @@ this:
+    |> sparens)
+    ("---", ("----", "---"))
+]}
+
```
This is out-of-scope of this tutorial, and I don't think introducing combinators to use them once is simpler for the target audience of this tutorial.

Also personally, I find the use of the left-to-right pipeline surprising in this context and  I would propose you to write
```ocaml
let string s ppf = Format.pp_print_string ppf s
let parens pp = Format.dprintf "@[(@[%t@]@,)@]" pp
let pair x y ppf = x ppf; y ppf

let () =
   Format.printf "%t" @@
  parens @@
    pair
      (string "---")
      (parens
         (pair
            (string "----")
            (parens @@ string "---")
         )
      )
```

edwintorok replied:
I'll remove it when I update the PR.

Octachron commented on api_docgen/Format_tutorial.mld:
```diff
@@ -247,6 +247,20 @@ break hint that precedes a closing parenthesis can show the boxes structure, if
+
+However the closing parenthesis don't line up with the opening ones, to
+correctly show the structure we need to introduce additional structural
+boxes after the parenthesis.
```
This is not an explanation of what it is happening: the issue is that the closing parenthesis is indented with respect to the indentation of box containing the break hints.

If one want to align the closing parenthesis with the opening one, there are two options:
* cancel the box indentation with a negative indentation
* add one more level of parent box with no indentation

edwintorok replied:
I've included something similar in `doc(format): mention 2 alternatives for getting structural boxes that line up vertically`, and added you as `Suggested-by` on the commit.

Octachron commented:
Note that I think it would be a good idea to have another tutorial for `Format` which spend more time on how to use `Format` rather than trying to describe what are breaks and boxes.

edwintorok commented:
> I am not convinced that using break hints with a negative indentation is not a better solution

The linked discussion gave the impression that negative indentation break hints should be avoided "You almost never need negative indentation with format. If you do, it’s more likely than your boxes are in the wrong places.".
Although looking at the compiler itself there are plenty of places where it is used.

If negative indentation isn't something to be discouraged, then perhaps both approaches should be shown, and let the developer pick the right approach for each situation.
What do you think?

> Note that I think it would be a good idea to have another tutorial for `Format` which spend more time on how to use `Format` rather than trying to describe what are breaks and boxes.

I agree, showing some examples on where to place boxes and break hints for common tasks like printing (recursive) data structures would be useful.
Although not a tutorial, the `fmt` library documentation shows the equivalent format strings for each combinator, it could be a starting point for a tutorial.
There are lot of other practicalities that aren't really covered (setting margin based on terminal size, use of unicode, etc.).

Octachron commented:
My opinion is that it is fine to use negative indentations in a controlled way in the code which is responsible for opening the box. For instance, with
```ocaml
"@[<v 2>(@, .... @;<0 -2>)@]"
```
the reader can infer that we are using a `2` indented box, except for the final `)` where we are restoring the alignment with the box. Contrarily, using negative indentation to fight against the alignment of parent boxes is not compositional and should be avoided.

edwintorok commented:
> where we are restoring the alignment with the box. Contrarily, using negative indentation to fight against the alignment of parent boxes is not compositional

Good point, so one negative indent hint towards the end with an exact match of the opening indent is probably fine (as long as you know for sure how big the opening indent is).

I'll rethink my PR and update it with this in mind, just changing the definition of `b` (instead of `print_cut`) should do it.

(We could try to define a rule of thumb about summing up all the negative indentations in a box and have an upper bound on that, but I think that if you'd find the need to decrease or increase the indentation gradually or multiple times within the same box, then it'd be better to open another box)

edwintorok committed:
doc(format): mention 2 alternatives for getting structural boxes that line up vertically

Controlled use of negative indentation (matching opening indentation) is
fine.
Another alternative is to use boxes with 0 indentation, mention both and
let the reader choose.

Suggested-by: Florian Angeletti <florian.angeletti@inria.fr>
Signed-off-by: Edwin Török <edwin@etorok.net>

edwintorok committed:
doc(format): add note about controlled use of negative indentation hints

Using a negative indentation hint that exceeds the opening indentation
would not be compositional.

We could try to come up with a more complicated rule about keeping the
running sum of indentation in a box >= 0, but if indentation needs to be
adjusted in places other than when opening and closing the box, then
opening another box might be better instead.
An one-off, exact match of negative and positive indentations
is a lot easier to understand and review in code.

Suggested-by: Florian Angeletti <florian.angeletti@inria.fr>
Signed-off-by: Edwin Török <edwin@etorok.net>

Octachron commented on api_docgen/Format_tutorial.mld:
```diff
@@ -265,6 +283,10 @@ When writing a pretty-printing routine, follow these simple rules:
++ Break hints with negative indentation should only be used when you are sure
+that you don't exceed the indentation of the opening box (e.g. both part of
+the same function and an exact match). Exceeding the indentation of the opening
+box would not be compositional.
```
I would be firmer here since we are in the midst of an imperative sentence: `would not be compositional` ⇒ `is not compositional`.

Octachron commented on api_docgen/Format_tutorial.mld:
```diff
@@ -247,6 +247,24 @@ break hint that precedes a closing parenthesis can show the boxes structure, if
+ )
+]}
+
+We can line up the opening and closing parenthesis vertically in two ways:
```
We might want to point out that the closing parenthesis is too far on the right due to the indentation before explaining how to avoid this misalignment.

edwintorok replied:
I've inserted a new commit ` doc(format): explain why the parenthesis are not lined up vertically`

>Each break hint increases indentation by one, both for opening and closing
>parenthesis, because the parenthesis are just characters and don't have special
>meaning for the formatter.
>Notice that the closing parenthesis is not vertically aligned with the opening parenthesis.

Maybe that repeats information that should already be obvious (the formatter only knows about boxes and break hints, so of course it makes no distinction between opening and closing parenthesis, they're just characters), but we're in the tutorial here, and explaining why you got the output might help the reader "connect the dots" on how Format works.

I've also added another commit at the beginning:
`doc(format): the example uses 'open_box 1',  `open_hovbox 1'`

It is quite obvious that the box needs to have an indentation of 1, otherwise the opening parenthesis wouldn't be indented either, but make this explicit

edwintorok committed:
doc(format): explain why the parenthesis are not lined up vertically

Signed-off-by: Edwin Török <edwin@etorok.net>

edwintorok committed:
doc(format): the example uses `open_box 1`, `open_hovbox 1`

`open_box 0` would not indent the opening parenthesis either.

Signed-off-by: Edwin Török <edwin@etorok.net>

Octachron commented on Changes:
```diff
@@ -79,6 +79,9 @@ _______________
 - #13045: Emphasize caution about behaviour of custom block finalizers.
   (Nick Barnes)
 
+- #13078: Fix Format tutorial on structural boxes to match the output.
```
The PR is now more complete than this description, maybe `update Format tutorial on structural boxes to mention alignment questions`?

Octachron reviewed (approved):
The current improved explanation sounds good to me, thanks !

Octachron commented on Changes:
```diff
@@ -79,6 +79,9 @@ _______________
   (Nick Barnes)
 
+- #13078: update Format tutorial on structural boxes to mention alignment
+ questions. (Edwin Török, review by Florian Angeletti)
```
The author and reviewer part should be on its own line (this helps changelog analysis tools).

edwintorok replied:
Thanks, I've updated the changelog and fixed the hygiene checks about line length.

edwintorok commented:
Thanks for the reviews.
Do you know if I'll need to update https://github.com/ocaml/ocaml.org/blob/main/data/tutorials/guides/0tt_00_formatting_text.md too once this is merged, or is there some automatic process that will eventually do that?

Octachron commented:
I am afraid that this is a different tutorial evolved from the same source, so it will need to be updated separately.
